libp2p/protocols/discovery/rendezvous.nim

```diff
@@ -0,0 +1,348 @@
+  conn.sendRegisterResponse(ttl)
+
+proc unregister(rdv: RendezVous, conn: Connection, u: Unregister) =
+  let nsSalted = u.ns.get("") & rdv.salt
+  try:
+    for index in rdv.indexes[nsSalted]:
+      if rdv.registered[index].peerId == conn.peerId:
+        rdv.registered[index].expireDate = rdv.defaultDT
+  except KeyError:
+    return
+
+proc discover(rdv: RendezVous, conn: Connection, d: Discover): Future[void] =
+  let ns = d.ns.get("")
+  if ns.len notin 0..255:
+    return conn.sendDiscoverResponseError(InvalidNamespace)
+  var
+    limit = min(DiscoverLimit, d.limit.get(DiscoverLimit))
+  let
+    nsSalted = ns & rdv.salt
+    cookie = if d.cookie.isSome():
+      try: Protobuf.decode(d.cookie.get(), type(Cookie))
+      except ProtobufReadError: return conn.sendDiscoverResponseError(InvalidCookie)
+    else: Cookie(offset: rdv.registered.low().uint64, ns: ns)
+    indexes = if ns != "":
+        try: rdv.indexes[nsSalted].s
+        except KeyError: return conn.sendDiscoverResponseError(InvalidNamespace)
+      else: toSeq(rdv.registered.low()..rdv.registered.high())
+  if indexes.len() == 0:
+    return conn.sendDiscoverResponse(@[], Cookie())
+  var offset = indexes[^1]
+  let s = collect(newSeq()):
+      for index in indexes:
+        if index.uint64 < cookie.offset: continue
+        if limit == 0:
+          offset = index
+          break
+        limit.dec()
+        rdv.registered[index].data
+  conn.sendDiscoverResponse(s, Cookie(offset: offset.uint64, ns: ns))
+
+proc advertisePeer(rdv: RendezVous, peer: PeerId, msg: seq[byte]) {.async.} =
+  let conn = await rdv.switch.dial(peer, RendezVousCodec)
+  defer: await conn.close()
+  await conn.writeLp(msg).wait(chronos.milliseconds(2500))
+  let
+    buf = await conn.readLp(4096).wait(chronos.milliseconds(2500))
+    msgRecv = Protobuf.decode(buf, type(Message))
+  if msgRecv.msgType.isNone() or msgRecv.msgType.get() != MessageType.RegisterResponse:
+    trace "Unexpected register response", peer, msgType = msgRecv.msgType
+  elif msgRecv.registerResponse.isNone() or
+       msgRecv.registerResponse.get().status.isNone() or
+       msgRecv.registerResponse.get().status.get() != ResponseStatus.Ok:
+    trace "Refuse to register", peer, response = msgRecv.registerResponse
+  rdv.sema.release()
+
+proc advertise*(rdv: RendezVous,
+                ns: string,
+                ttl: uint64 = MinimumTTL) {.async.} =
+  let sprBuff = rdv.switch.peerInfo.signedPeerRecord.envelope.encode()
+  if sprBuff.isErr() or ns.len() notin 1..255 or ttl notin MinimumTTL..MaximumTTL:
+    return
+  let
+    r = Register(ns: some(ns), signedPeerRecord: some(sprBuff.get()), ttl: some(ttl))
+    msg = Protobuf.encode(Message(msgType: some(MessageType.Register), register: some(r)))
+  rdv.save(ns, rdv.switch.peerInfo.peerId, r, ttl)
+  for peer in rdv.peers:
+    try:
+      await rdv.sema.acquire()
+      asyncSpawn rdv.advertisePeer(peer, msg)
+    except CancelledError as exc:
+      rdv.sema.release()
```

Comment:
That's not correct, the only point where we can get cancelled / have exception here is in the acquire (asyncSpawn doesn't block)